### ISSUE TITLE
no decode delegate

### ISSUE THREAD

**Issue by dolejm1** [JuliaIO/ImageMagick.jl#25]
HI,
I have problem to load jpg or png image by Images.load.
The error is:
WARNING: ErrorException("no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/501")
 in error at error.jl:21
ERROR: no decode delegate for this image format`JPEG' @ error/constitute.c/ReadImage/501

and similarly for png. This is ImageMagick error I guess, but my ImageMagick says:
convert -version
Version: ImageMagick 6.9.3-2 Q16 x64 2016-01-23 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2015 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib cairo freetype jng jp2 jpeg lcms lqr openexr pangocairo png ps rsvg tiff webp xml zlib

so there is a delegate for jpeg and png files. The problem appeared after installing julia 4.3.
I've installed the last ImageMagick, updated all the julia packages, and build everything without errors.

**Comment by ajsteven130** [comment]
I am getting exactly the same thing using the 32 bit builds of all of the above on Windows 7.  

My julia version is 0.4.3 installed to the standard location from the officially released self-extracting archives. All packages were added using `Pkg.add()` and built without errors.

ImageMagick was installed according to the directions in the README for this repository and I was sure to include the c and c++ libraries and headers.

Complete imagemagick info:

```
convert -version
Version: ImageMagick 6.9.3-5 Q16 x86 2016-02-20 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2015 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC Modules OpenMP 
Delegates (built-in): bzlib cairo freetype jng jp2 jpeg lcms lqr openexr pangocairo png ps rsvg tiff webp xml zlib
```

ouput from julia is:

```
julia> using Images
julia> im = load("file.jpg")
WARNING: ErrorException("no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/501")
 in error at error.jl:21
 in error at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\libmagickwand.jl:149
 in load_ at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\ImageMagick.jl:74
ERROR: no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/501
 in error at error.jl:21
 in error at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\libmagickwand.jl:149
 in load_ at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\ImageMagick.jl:74
```

The same file, `file.jpg`, loads and can be manipulated normally from the command line with ImageMagick commands like `convert`.

I

**Comment by SimonDanisch** [comment]
Hi, sorry for the inconvenience... 
Can you post the contents of `ImageMagick\deps\deps.jl` ?

**Comment by ajsteven130** [comment]
Hi, thanks very much for all the hard work and the quick response. The contents of my `Ìmagemagick\deps\deps.jl` file is:

```
# This is an auto-generated file; do not edit

# Pre-hooks

# Macro to load a library
macro checked_lib(libname, path)
    ((VERSION >= v"0.4.0-dev+3844" ? Base.Libdl.dlopen_e : Base.dlopen_e)(path) == C_NULL) && error("Unable to load \n\n$libname ($path)\n\nPlease re-run Pkg.build(package), and restart Julia.")
    quote const $(esc(libname)) = $path end
end

# Load dependencies
@checked_lib libwand "C:\\Program Files\\ImageMagick-6.9.3-Q16\\CORE_RL_wand_.DLL"

# Load-hooks

function init_deps()
    ENV["MAGICK_CONFIGURE_PATH"]    = "C:\\Users\\A6419643\\.julia\\v0.4\\ImageMagick\\deps\\usr\\lib\\x86"
    ENV["MAGICK_CODER_MODULE_PATH"] = "C:\\Users\\A6419643\\.julia\\v0.4\\ImageMagick\\deps\\usr\\lib\\x86"
end
init_deps()
```

Something that jumps out at me immediately is that `"C:\\Users\\A6419643\\.julia\\v0.4\\ImageMagick\\deps\\usr\\lib\\x86"` does not exist.

**Comment by SimonDanisch** [comment]
Hah, good to know! Can you set both paths to `C:\\Program Files\\ImageMagick-6.9.3-Q16\\` manually? If that works, fixing the buld script should be straight forward :)

**Comment by ajsteven130** [comment]
I tried setting both paths to `C:\\Program Files\\ImageMagick-6.9.3-Q16\\` but I continue to get the same error. I checked in the REPL to be sure both values were properly set after ImageMagick recompiled. They both remained set to the hand coded values: `C:\\Program Files\\ImageMagick-6.9.3-Q16\\`.

I tried running `Pkg.build("ImageMagick")` and this resulted in overwriting the manually entered values (as you would expect);

Here is the code from the REPL session:

```
C:\Users\A6419643
λ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.3 (2016-01-12 21:37 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  i686-w64-mingw32

julia> using Images

shell> cd Desktop\\
C:\Users\A6419643\Desktop

julia> image = load("loic.jpg")
WARNING: ErrorException("no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/501")
 in error at error.jl:21
 in error at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\libmagickwand.jl:149
 in load_ at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\ImageMagick.jl:74
ERROR: no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/501
 in error at error.jl:21
 in error at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\libmagickwand.jl:149
 in load_ at C:\Users\A6419643\.julia\v0.4\ImageMagick\src\ImageMagick.jl:74

julia> ENV["MAGICK_CONFIGURE_PATH"]
"C:\\Program Files\\ImageMagick-6.9.3-Q16\\"

julia> ENV["MAGICK_CODER_MODULE_PATH"]
"C:\\Program Files\\ImageMagick-6.9.3-Q16\\"

julia>
```

**Comment by SimonDanisch** [comment]
ugh... `ImageMagick` is such a pain to set up correctly -.- .
So maybe we DO need a call to `ccall((:MagickWandGenesis,libwand), Void, ())` after all?

So try:

``` Julia
# This is an auto-generated file; do not edit

# Pre-hooks

# Macro to load a library
macro checked_lib(libname, path)
    ((VERSION >= v"0.4.0-dev+3844" ? Base.Libdl.dlopen_e : Base.dlopen_e)(path) == C_NULL) && error("Unable to load \n\n$libname ($path)\n\nPlease re-run Pkg.build(package), and restart Julia.")
    quote const $(esc(libname)) = $path end
end

# Load dependencies
@checked_lib libwand "C:\\Program Files\\ImageMagick-6.9.3-Q16\\CORE_RL_wand_.DLL"

# Load-hooks

function init_deps()
    ENV["MAGICK_CONFIGURE_PATH"]    = "C:\\Program Files\\ImageMagick-6.9.3-Q16"
    ENV["MAGICK_CODER_MODULE_PATH"] = "C:\\Program Files\\ImageMagick-6.9.3-Q16"
ccall((:MagickWandGenesis,libwand), Void, ())
end
init_deps()
```

Sorry for this fuzzy stuff, I have no idea how to do this correctly and there doesn't seem any congruent documentation for this.
See: https://github.com/JuliaIO/ImageMagick.jl/issues/7

**Comment by ajsteven130** [comment]
That fixed it! Thank you! I hope this means that fixing up the build script is still relatively easy.

In the mean time, I hacked together an approach for loading images via python's scikit image that you can see [here](https://gist.github.com/ajsteven130/6da424a7e4478a8bd07f).

Do you see any value in working that solution out a little more fully to have alongside imagemagick? At this point it is rather hacky, but I got it working while debugging the ImageMagick side of things with you.

**Comment by SimonDanisch** [comment]
Hm I'm not sure... at some point we took out the call to `MagickWandGenesis` for a reason... I can't clearly remember the reason anymore, though :D I see if I can reproduce the behaviour and try it out myself.

> In the mean time, I hacked together an approach for loading images via python's scikit image that you can see here.

Great! `FileIO` supports different loader/saver libraries and can use them as a fallback. So if there are more IO libraries, it should be less problematic if one fails! You can look at the `FileIO` documentation and if that's unclear don't hesitate to open an issue about missing documentation!

**Comment by zorg24** [comment]
Just so you know I was actually having the same issue, but I got it working without the `ccall((:MagickWandGenesis,libwand), Void, ())` line, I only had to change the two ENV lines.

**Comment by ajsteven130** [comment]
@zorg24 - it is entirely possible that I managed to screw something up manually setting the variables, but checking their values in the REPL convinced me that I hadn't messed up too badly. Perhaps the difference in our experiences is one reason why the history of that line is unclear for @SimonDanisch .

@SimonDanisch Thanks a ton for all your quick responses. I'll take a look at the `FileIO` documentation and try to put something together using PyCall and scikit image.

It seems like a good long term goal would be to start getting some of the decoding methods put into native Julia, but it seems like a daunting task after looking at [imread](https://github.com/luispedro/imread)

**Comment by SimonDanisch** [comment]
From the C examples I can tell, that it rather looks like everything you do with imagemagick should be wrapped in:

``` Julia
MagickWandGenesis()
...
MagickWandTerminus()
```

This calls for:

``` Julia
function __init__()
MagickWandGenesis()
atexit(MagickWandTerminus)
end
```

I'll try out if this works as expected!

**Comment by ajsteven130** [comment]
Where exactly should the new **init**() function you propose go? in `libmagickwand.jl` I have:

```
# Initialize the library
function __init__()
    init_deps()
    !have_imagemagick && warn("ImageMagick utilities not found. Install for more file format support.")
end
```

should it replace this function?

Sorry for the confusion, I am just getting up to speed with julia. Thanks again for your help and patience.

**Comment by SimonDanisch** [comment]
The confusion is deeply embedded into this library :D But yeah, that'd be the one that needs tweaking.

I'm wondering, if we started using different initialization for different operating systems for the wrong reasons. In the C examples, everything looks pretty straightforward. I'll play around with a new build script and see how it goes!

**Comment by ajsteven130** [comment]
ok - let me know if you need someone to test on windows. I'll wait to hear something from you. 

Thanks again for your help with this.

**Comment by dognotdog** [comment]
FWIW, I ran into this issue today with Julia 0.4.5 on Windows 7, but here's a quick fix for `deps.jl`:

If your ImageMagick install is at `C:\\Program Files\\ImageMagick-6.9.3-Q16` set

```
ENV["MAGICK_CODER_MODULE_PATH"] = "C:\\Program Files\\ImageMagick-6.9.3-Q16\\modules\\coders"
```

in `init_deps()` and that seems to fix it.

**Comment by SimonDanisch** [comment]
does https://github.com/JuliaIO/ImageMagick.jl/pull/28 also work for you?

**Comment by dognotdog** [comment]
@SimonDanisch nope, haven't tried that, I only discovered this issue after applying my hackish fix. This must be some regression, too, as it used to work with Julia 0.4.0 and 0.3.x.

**Comment by SimonDanisch** [comment]
Okay...Well this line: https://github.com/JuliaIO/ImageMagick.jl/pull/28/files#diff-95f0ceb18109e984f0198d318b09b4ebR63 looks like it already does what you describe :)
So I guess we can assume that the PR fixes it.

**Comment by ghost** [comment]
I ran into a similar issue on MacOS X 10.9.5 today. Any idea how to fix it? I ran Pkg.add("ImageMagick"), Pkg.update(), and restarted Julia.

> julia> using ImageMagick

> julia> @gif for i=1:100
           plot(x,x.^2+.01*i)
       end every 10
convert: no decode delegate for this image format `PNG' @ error/constitute.c/ReadImage/504.
convert: no images defined `/var/folders/q1/4bt1n5dn43n7cyw5pthmj8k40000gp/T/julia2peMzf.gif' @ error/convert.c/ConvertImageCommand/3258.
WARNING: Tried to create gif using convert (ImageMagick), but got error: ErrorException("failed process: Process(`convert -delay 5 -loop 0 /var/folders/q1/4bt1n5dn43n7cyw5pthmj8k40000gp/T/tmpFN53ii/*.png -alpha off /var/folders/q1/4bt1n5dn43n7cyw5pthmj8k40000gp/T/julia2peMzf.gif`, ProcessExited(1)) [1]")
ImageMagick can be installed by executing `Pkg.add("ImageMagick")`
Will try ffmpeg, but it's lower quality...)
ERROR: could not spawn `ffmpeg -v 0 -framerate 20 -loop 0 -i /var/folders/q1/4bt1n5dn43n7cyw5pthmj8k40000gp/T/tmpFN53ii/%06d.png -y /var/folders/q1/4bt1n5dn43n7cyw5pthmj8k40000gp/T/julia2peMzf.gif`: no such file or directory (ENOENT)
 in _jl_spawn(::String, ::Array{String,1}, ::Ptr{Void}, ::Base.Process, ::RawFD, ::RawFD, ::RawFD) at ./process.jl:321
 in #424 at ./process.jl:478 [inlined]
 in setup_stdio(::Base.##424#425{Cmd,Ptr{Void},Base.Process}, ::Tuple{RawFD,RawFD,RawFD}) at ./process.jl:466
 in #spawn#423(::Nullable{Base.ProcessChain}, ::Function, ::Cmd, ::Tuple{RawFD,RawFD,RawFD}, ::Bool, ::Bool) at ./process.jl:477
 in #spawn#423(::Nullable{Base.ProcessChain}, ::Function, ::Cmd, ::Tuple{RawFD,RawFD,RawFD}, ::Bool, ::Bool) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in run(::Cmd) at ./process.jl:591
 in #buildanimation#348(::Int64, ::Int64, ::Function, ::String, ::String) at /Users/sebastianament/.julia/v0.5/Plots/src/animation.jl:87
 in #gif#345(::Array{Any,1}, ::Function, ::Plots.Animation, ::String) at /Users/sebastianament/.julia/v0.5/Plots/src/animation.jl:58
 in gif(::Plots.Animation) at /Users/sebastianament/.julia/v0.5/Plots/src/animation.jl:58
 in macro expansion; at /Users/sebastianament/.julia/v0.5/Plots/src/animation.jl:155 [inlined]
 in anonymous at ./<missing>:?

**Comment by timholy** [comment]
Did you already try the troubleshooting steps posted in the README?

**Comment by Allardvm** [comment]
This should be fixed by https://github.com/JuliaIO/ImageMagick.jl/commit/b0bc710b36608ec5c759f6babe9252794a5e3348. Could you try with release 0.4?

**Comment by SimonDanisch** [comment]
Should be fixed by #121! Please try again on master!
